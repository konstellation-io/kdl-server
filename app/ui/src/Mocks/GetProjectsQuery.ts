import { project1, project2, projectNoAccess, projectArchived } from './entities/project';

export default {
  projects: [project1, project2, projectNoAccess, projectArchived],
};
