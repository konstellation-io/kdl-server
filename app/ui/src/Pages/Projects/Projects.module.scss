@import 'Styles/variables';

.container {
  display: flex;
  flex-wrap: wrap;
  padding: 4 * $grid-unit 4.5 * $grid-unit 4 * $grid-unit 5 * $grid-unit;
  overflow: auto;
}

.adminsModal > div:last-child > div > div:last-child {
  display: none;
}
