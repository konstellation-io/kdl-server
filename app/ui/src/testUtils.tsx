import * as React from 'react';
import { ApolloClient, ApolloProvider } from '@apollo/client';
import { render } from '@testing-library/react';
import { mount } from 'enzyme';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export async function apolloRender(component: JSX.Element, client: ApolloClient<any>) {
  render(<ApolloProvider client={client}>{component}</ApolloProvider>);
}

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
export const HookWrapper = ({ hook }: any) => <div hook={hook()} />; // eslint-disable-line @typescript-eslint/no-explicit-any
export async function apolloHookRender(
  // eslint-disable-next-line @typescript-eslint/ban-types
  hook: Function,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  client: ApolloClient<any>,
) {
  const wrapper = mount(
    <ApolloProvider client={client}>
      <HookWrapper hook={hook} />
    </ApolloProvider>,
  );

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return wrapper.find('div').props().hook;
}

export const loadingHandler = () =>
  new Promise(() => {
    // This is intentional
  });
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const dataHandler = (data: any) => () => Promise.resolve({ data });
export const errorHandler = (error: string) => () => Promise.reject({ error });

export function getSnapshot() {
  return document.querySelector('body');
}

export const ERROR_MESSAGE = 'Something went wrong, refresh the page or try again later.';
