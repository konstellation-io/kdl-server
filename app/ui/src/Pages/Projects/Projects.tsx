import { ErrorMessage, ModalContainer, ModalLayoutConfirmList, SpinnerCircular } from 'kwc';
import AdminEmail from './components/AdminEmail/AdminEmail';

import AddProject from './components/Project/AddProject';
import { GetProjects } from 'Graphql/queries/types/GetProjects';
import Project from './components/Project/Project';
import ProjectsBar from 'Components/ProjectsBar/ProjectsBar';
import React, { useState } from 'react';
import styles from './Projects.module.scss';
import useProjectFilters from 'Graphql/client/hooks/useProjectFilters';
import { useQuery, useReactiveVar } from '@apollo/client';

import GetProjectsQuery from 'Graphql/queries/getProjects';
import { projectFilters } from 'Graphql/client/cache';

export type ProjectAdmins = {
  projectName: string;
  administrators: string[];
};

function Projects() {
  const [showProjectAdmins, setShowProjectAdmins] = useState<ProjectAdmins | null>(null);

  const { data, error, loading } = useQuery<GetProjects>(GetProjectsQuery);
  const filters = useReactiveVar(projectFilters);

  const { filterProjects, sortProjects } = useProjectFilters();
  if (loading) return <SpinnerCircular />;
  if (error || !data) return <ErrorMessage />;

  let projects = filterProjects(data.projects, filters);
  projects = sortProjects(projects, filters.order);

  return (
    <>
      <ProjectsBar />
      <div className={styles.container}>
        {[
          ...projects.map((project) => (
            <Project key={project.id} project={project} showAdmins={setShowProjectAdmins} />
          )),
          <AddProject key="add-project" />,
        ]}
      </div>
      {showProjectAdmins !== null && (
        <ModalContainer
          title={showProjectAdmins.projectName}
          actionButtonLabel="Close"
          actionButtonCancel="Cancel"
          onAccept={() => setShowProjectAdmins(null)}
          className={styles.adminsModal}
          blocking
        >
          <ModalLayoutConfirmList message="You can email the following users to ask for access to this project">
            {showProjectAdmins.administrators.map((email) => (
              <AdminEmail email={email} key={email} />
            ))}
          </ModalLayoutConfirmList>
        </ModalContainer>
      )}
    </>
  );
}

export default Projects;
