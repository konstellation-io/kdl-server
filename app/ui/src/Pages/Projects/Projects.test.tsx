import * as React from 'react';
import Projects from './Projects';
import GetProjectsQuery from 'Graphql/queries/getProjects';
import { createMockClient } from 'mock-apollo-client';
import { screen, fireEvent, cleanup, within } from '@testing-library/react';
import { apolloRender, loadingHandler, dataHandler, errorHandler, getSnapshot, ERROR_MESSAGE } from 'testUtils';
import data from 'Mocks/GetProjectsQuery';
import { StaticRouter } from 'react-router';

const Component = (
  <StaticRouter>
    <Projects />
  </StaticRouter>
);

let mockClient;
beforeEach(() => {
  mockClient = createMockClient();
});
afterEach(cleanup);

describe('When component is loading', () => {
  beforeEach(() => {
    mockClient.setRequestHandler(GetProjectsQuery, loadingHandler);
    apolloRender(Component, mockClient);
  });

  it('should render without crashing', () => {
    expect(screen.getByTestId('spinner')).toBeInTheDocument();
  });
});

describe('When data is ready', () => {
  beforeEach(() => {
    mockClient.setRequestHandler(GetProjectsQuery, dataHandler(data));
    apolloRender(Component, mockClient);
  });

  it('should render without crashing', () => {
    expect(getSnapshot()).toMatchSnapshot();
  });

  it('should show two projects and one add project', function () {
    // Arrange.
    const projects = screen.getAllByTestId('project');
    const newProject = screen.getByTestId('add-project');

    // Assert.
    expect(projects.length).toBe(2);
    expect(newProject).toBeInTheDocument();
  });

  test('when filtering by ARCHIVED, it should only show one project', () => {
    // Act.
    // Opens filter
    const projectsFilter = screen.getByText('active', { exact: false });
    fireEvent.click(projectsFilter);

    // Select Archived projects
    const archivedOption = screen.getByText('archived', { exact: false });
    fireEvent.click(archivedOption);

    // Assert.
    const projects = screen.getAllByTestId('project');
    const archivedProject = within(projects[0]).getByText('Archived');
    expect(projects.length).toBe(1);
    expect(archivedProject).toBeInTheDocument();
  });

  test('when filtering by INACCESSIBLE, it should only show one project', () => {
    // Act.
    // Opens filter
    const projectsFilter = screen.getByText('active', { exact: false });
    fireEvent.click(projectsFilter);

    // Select Inaccessible projects
    const inaccessibleOption = screen.getByText('inaccessible', {
      exact: false,
    });
    fireEvent.click(inaccessibleOption);

    // Assert.
    const projects = screen.getAllByTestId('project');
    const inaccessibleProject = within(projects[0]).getByText('No Access');
    expect(projects.length).toBe(1);
    expect(inaccessibleProject).toBeInTheDocument();
  });

  test('when filtering by All, it should four projects', () => {
    // Act.
    // Opens filter
    const projectsFilter = screen.getByText('active', { exact: false });
    fireEvent.click(projectsFilter);

    // Select All projects
    const allOption = screen.getByText('all', { exact: false });
    fireEvent.click(allOption);

    // Assert.
    const projects = screen.getAllByTestId('project');
    expect(projects.length).toBe(4);
  });
});

describe('When there is an error', () => {
  beforeEach(() => {
    mockClient.setRequestHandler(GetProjectsQuery, errorHandler(''));
    apolloRender(Component, mockClient);
  });

  it('should render without crashing', () => {
    expect(screen.getByText(ERROR_MESSAGE)).toBeInTheDocument();
  });
});

// Arrange.
// Act.
// Assert.
