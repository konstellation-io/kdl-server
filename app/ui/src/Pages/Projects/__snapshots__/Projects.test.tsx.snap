// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When data is ready should render without crashing 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <div
        class="_pcC7V left"
      >
        <div
          class="container"
          data-testid="filterProjects"
        >
          <div
            class="formSelect _2rBeV _1z7Bn"
          >

            <div
              class="_Z0bIQ"
            >
              <div
                class="_3Loey"
                style="height: 67px; min-height: 67px;"
              >
                Active (2)
              </div>
              <div
                aria-hidden="true"
                class="rah-static rah-static--height-zero _3bqns"
                style="height: 0px; overflow: hidden;"
              >
                <div
                  style="display: none;"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="container"
          data-testid="sortProjects"
        >
          <p
            class="label"
          >
            SORT BY
          </p>
          <div
            class="_2rBeV _1z7Bn"
          >

            <div
              class="_Z0bIQ"
            >
              <div
                class="_3Loey"
                style="height: 40px; min-height: 40px;"
              >
                Creation date
              </div>
              <div
                aria-hidden="true"
                class="rah-static rah-static--height-zero _3bqns"
                style="height: 0px; overflow: hidden;"
              >
                <div
                  style="display: none;"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="_2u-ZX"
      >
        <a
          data-testid="buttonLink"
          href="/new-project"
        >
          <div
            class="addProjectButton _1lqJX _3kQFd _zeH0N undefined"
            style="height: 40px; line-height: 40px;"
            tabindex="-1"
            title=""
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root icon-small"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
            <span>
              Add project
            </span>
          </div>
        </a>
      </div>
    </div>
    <div
      class="container"
    >
      <a
        href="/projects/projectId1"
      >
        <div
          class="container"
          data-testid="project"
        >
          <div
            class="sup"
          >
            <div
              class="bg"
            >
              <div
                class="bgBand"
              />
            </div>
            <div
              class="content"
            >
              <p
                class="name"
                data-testid="projectName"
              >
                projectName1
              </p>
            </div>
          </div>
          <div
            class="inf"
          >
            <div
              class="bg"
            >
              <div
                class="bgBand"
              />
            </div>
            <div
              class="content"
            >
              <p
                class="description"
                title="projectDescription1"
              >
                projectDescription1
              </p>
              <div
                class="creationDate"
              >
                <p
                  class="creationDateLabel"
                >
                  CREATED:
                </p>
                <p
                  class="date"
                >
                  Feb 2, 2020 12:00 AM
                </p>
              </div>
            </div>
          </div>
          <div
            class="band"
          >
            <div
              class="otherLabels"
            />
          </div>
          <div
            class="square"
          >
            <div
              class="repoType"
            >
              <div
                class="square out medium notAnimate"
                style="height: 38px; width: 38px;"
              >
                <div
                  class="s4"
                />
                <div
                  class="s3"
                />
                <div
                  class="s2"
                />
                <div
                  class="s1"
                />
              </div>
            </div>
          </div>
        </div>
      </a>
      <a
        href="/projects/projectId2"
      >
        <div
          class="container"
          data-testid="project"
        >
          <div
            class="sup"
          >
            <div
              class="bg"
            >
              <div
                class="bgBand"
              />
            </div>
            <div
              class="content"
            >
              <p
                class="name"
                data-testid="projectName"
              >
                projectName2
              </p>
            </div>
          </div>
          <div
            class="inf"
          >
            <div
              class="bg"
            >
              <div
                class="bgBand"
              />
            </div>
            <div
              class="content"
            >
              <p
                class="description"
                title="projectDescription2"
              >
                projectDescription2
              </p>
              <div
                class="creationDate"
              >
                <p
                  class="creationDateLabel"
                >
                  CREATED:
                </p>
                <p
                  class="date"
                >
                  Mar 3, 2020 12:00 AM
                </p>
              </div>
            </div>
          </div>
          <div
            class="band"
          >
            <div
              class="otherLabels"
            />
          </div>
          <div
            class="square"
          >
            <div
              class="repoType"
            >
              <div
                class="square out medium notAnimate"
                style="height: 38px; width: 38px;"
              >
                <div
                  class="s4"
                />
                <div
                  class="s3"
                />
                <div
                  class="s2"
                />
                <div
                  class="s1"
                />
              </div>
            </div>
          </div>
        </div>
      </a>
      <div
        class="addProjectContainer"
        data-testid="add-project"
      >
        <a
          href="/new-project"
        >
          <div
            class="addProjectBg"
          >
            <div
              class="addProjectBgUpper"
            />
            <div
              class="addProjectBgLower"
            />
          </div>
        </a>
        <div
          class="addProjectLabel"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root icon-regular"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
          <p>
            ADD PROJECT
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
